### PR TITLE
Add gulp master task

### PR DESCRIPTION
**Pull request by qinezh**

### CODE REVIEW

File: tools/Deployment/config_gulp.json [hunk 1]
```diff
@@ -15,5 +20,19 @@
   },
   "firefox": {
     "version": "46.0.1"
+  },
+  "choco": {
+    "homeDir": "../../../docfx/src/nuspec/chocolatey/docfx/",
```
Comment (superyyrrzz) [inline tools/Deployment/config_gulp.json]: >../docfx/ [](start = 22, length = 9)

remove? seems too many ../

File: tools/Deployment/gulpfile.js [hunk 2]
```diff
@@ -10,9 +10,15 @@ let del = require("del");
 let glob = require("glob");
 let gulp = require("gulp");
 let nconf = require("nconf");
-let spawn = require("child-process-promise").spawn;
 
-let configFile = path.join(__dirname, "config_gulp.json");
+let Common = require("./out/common").Common;
```
Comment (superyyrrzz) [inline tools/Deployment/gulpfile.js]: >out [](start = 24, length = 3)

read it from tsconfig.json, then path.join?

File: tools/Deployment/config_gulp.json [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
-    "home": "../../../docfx/",
+    "home": "../../",
+    "repoUrl": "git@github.com-ci:dotnet/docfx.git",
     "docfxSeedHome": "../../../docfx-seed/",
     "e2eTestsHome": "../../test/docfx.E2E.Tests/",
     "targetFolder": "../../target/",
     "artifactsFolder": "../../artifacts/",
-    "exe": "../../target/Release/docfx/docfx.exe"
+    "exe": "../../target/Release/docfx/docfx.exe",
+    "releaseNotePath": "../../RELEASENOTE.md",
```
Comment (hellosnow) [inline tools/Deployment/config_gulp.json]: >../.. [](start = 24, length = 5)

how about removing "../.." and contact with "home" path?

File: tools/Deployment/lib/common.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+export class Guard {
+    static argumentNotNull(argumentValue: Object, argumentName: string) {
+        if (argumentValue === null || argumentValue === undefined) {
+            throw new Error(`${argumentName} can't be null/undefined.`);
+        }
+    }
+
+    static argumentNotNullOrEmpty(stringValue: string, argumentName: string) {
+        if (stringValue === null || stringValue === "") {
```
Comment (hellosnow) [inline tools/Deployment/lib/common.ts]: >stringValue === null [](start = 12, length = 20)

need we check argumentValue === undefined or call  argumentNotNull directly?

File: tools/Deployment/lib/common.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+        fs.writeFileSync(targetPath, buffer);
+    }
+
+    static computeSha1FromZip(zipPath: string): string {
+        Guard.argumentNotNullOrEmpty(zipPath, "zipPath");
+
+        if (!zipPath) {
+            throw new Error(`${zipPath} can't null or undefined.`);
+        }
```
Comment (hellosnow) [inline tools/Deployment/lib/common.ts]: can remove since guard has already check

File: tools/Deployment/lib/common.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    static getVersionFromReleaseNote(releaseNotePath: string): string {
+        Guard.argumentNotNullOrEmpty(releaseNotePath, "releaseNotePath");
+
+        if (!fs.existsSync(releaseNotePath)) {
+            throw new Error(`${releaseNotePath} doesn't exist.`);
+        }
+
+        let regex = /\(Current\s+Version:\s+[vV]([\d\.]+)\)/;
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: > /\(Current\s+Version:\s+[vV]([\d\.]+)\)/ [](start = 19, length = 41)

//i to ignore case

File: tools/Deployment/lib/common.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        let content = fs.readFileSync(releaseNotePath, "utf8");
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >readFileSync [](start = 25, length = 12)

first line is enough

File: tools/Deployment/lib/common.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    static getDescriptionFromReleaseNote(releaseNotePath: string): string {
+        Guard.argumentNotNullOrEmpty(releaseNotePath, "releaseNotePath");
+
+        if (!fs.existsSync(releaseNotePath)) {
+            throw new Error(`${releaseNotePath} doesn't exist.`);
+        }
+
+        let regex = /---{3,}\r?\n([\s\S]+?)[vV][\d\.]+\r?\n---{3,}\r?\n/;
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >-- [](start = 21, length = 2)

remove

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >---{3,} [](start = 59, length = 7)

also here

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >regex [](start = 12, length = 5)

why should version wrapped by 5+ `-` ?

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >regex [](start = 12, length = 5)

the regex looks to match sth. similar to
```
-----

Hello v123
-----

```
why is the description wrapped by 5+ `-` and must append version number?

File: tools/Deployment/lib/common.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+        let match = regex.exec(content);
+        if (!match || match.length < 2) {
+            throw new Error(`Can't parse description from ${releaseNotePath} in current version part.`);
+        }
+
+        return match[1].trim();
+    }
+
+    static async isReleaseNoteUpdatedAsync(gitRootPath: string, releaseNotePath: string): Promise<boolean> {
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >isReleaseNoteUpdatedAsync [](start = 17, length = 25)

what if release note is changed however version number is not changed?

File: tools/Deployment/lib/common.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            return null;
+        }
+
+        return content.split(/\r?\n/);
+    }
+}
+
+export class Guard {
+    static argumentNotNull(argumentValue: Object, argumentName: string, message = null) {
+        if (argumentValue === null || argumentValue === undefined) {
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >undefined [](start = 56, length = 9)

to check undefined, use typeof
[http://stackoverflow.com/questions/3390396/how-to-check-for-undefined-in-javascript](http://stackoverflow.com/questions/3390396/how-to-check-for-undefined-in-javascript)

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >null [](start = 30, length = 4)

suggest reading [https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined)

Comment (qinezh) [inline tools/Deployment/lib/common.ts]: I think the guide is for those developing the TypeScript compiler, not a general statement about using null with TypeScript...

---
In reply to: [111542090](https://github.com/dotnet/docfx/pull/1542#discussion_r111542090) [](ancestors = 111542090)

Comment (qinezh) [inline tools/Deployment/lib/common.ts]: In ES2015, `undefined` can't be overwritten, so i think it's ok to use `===` to check it.

---
In reply to: [111541976](https://github.com/dotnet/docfx/pull/1542#discussion_r111541976) [](ancestors = 111541976)

**Comment by superyyrrzz** [comment]
:shipit:

**Comment by hellosnow** [comment]
:shipit:

File: tools/Deployment/lib/common.ts [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    static getDescriptionFromReleaseNote(releaseNotePath: string): string {
+        Guard.argumentNotNullOrEmpty(releaseNotePath, "releaseNotePath");
+
+        if (!fs.existsSync(releaseNotePath)) {
+            throw new Error(`${releaseNotePath} doesn't exist.`);
+        }
+
+        let regex = /\r?\n\s*v[\d\.]+\s*\r?\n-{3,}\r?\n([\s\S]+?)(\r?\n\s*v[\d\.]+\s*\r?\n-{3,}\r?\n|$)/i;
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >\r? [](start = 21, length = 3)

no need

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >\r?\n [](start = 95, length = 5)

no need?

Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >( [](start = 65, length = 1)

no need to group?

File: tools/Deployment/lib/common.ts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        return match[1].trim();
+    }
+
+    static async isReleaseNoteVersionChangedAsync(releaseNotePath: string): Promise<boolean> {
+        let versionFromTag = await this.getCurrentVersionFromGitTag();
+        let versionFromReleaseNote = this.getVersionFromReleaseNote(releaseNotePath);
+
+        return `v${versionFromTag}`.toLowerCase() !== versionFromReleaseNote.toLowerCase();
```
Comment (vicancy) [inline tools/Deployment/lib/common.ts]: >versionFromTag [](start = 19, length = 14)

tag already contains `v` ?

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**